pkg/scheduler/framework/plugins/volumebinding/volume_binding.go

```diff
@@ -46,9 +70,141 @@ func (pl *VolumeBinding) Filter(cs *framework.CycleState, pod *v1.Pod, nodeInfo
+	}
+	if allBound {
+		// no need to bind volumes
+		return nil
+	}
+	klog.V(5).Infof("Trying to bind volumes for pod \"%v/%v\"", pod.Namespace, pod.Name)
+	err = pl.volumeBinder.Binder.BindPodVolumes(pod)
+	if err != nil {
+		klog.V(1).Infof("Failed to bind volumes for pod \"%v/%v\": %v", pod.Namespace, pod.Name, err)
+		return migration.ErrorToFrameworkStatus(err)
+	}
+	klog.V(5).Infof("Success binding volumes for pod \"%v/%v\"", pod.Namespace, pod.Name)
+	return nil
+}
+
+// Unreserve is called by the scheduling framework when a reserved pod was
+// rejected in a later phase.
+// TODO(cofyc) Revert assumed PV/PVC cache, see http://issues.k8s.io/82934#issuecomment-538269188
+func (pl *VolumeBinding) Unreserve(cs *framework.CycleState, pod *v1.Pod, nodeName string) {
+	pl.volumeBinder.DeletePodBindings(pod)
+	return
+}
+
+// PostBind is called after a pod is successfully bound.
+func (pl *VolumeBinding) PostBind(cs *framework.CycleState, pod *v1.Pod, nodeName string) {
+	pl.volumeBinder.DeletePodBindings(pod)
+	return
+}
+
+// Args represents volume binder arguments.
+type Args struct {
+	BindTimeoutSeconds *int64 `json:"bindTimeoutSeconds,omitempty"`
+}
+
+// New initializes a new plugin and returns it.
+func New(config *runtime.Unknown, handle framework.FrameworkHandle) (framework.Plugin, error) {
+	nodeInformer := handle.SharedInformerFactory().Core().V1().Nodes()
+	pvcInformer := handle.SharedInformerFactory().Core().V1().PersistentVolumeClaims()
+	pvInformer := handle.SharedInformerFactory().Core().V1().PersistentVolumes()
+	storageClassInformer := handle.SharedInformerFactory().Storage().V1().StorageClasses()
+	var args Args
+	if err := framework.DecodeInto(config, &args); err != nil {
+		return nil, err
+	}
+	// args.BindTimeoutSeconds is derived from schedulerOptions.bindTimeoutSeconds which has default value set
+	if args.BindTimeoutSeconds == nil {
+		return nil, fmt.Errorf("failed to initialize %s plugin, expects bindTimeoutSeconds not to be nil", Name)
+	}
+	volumeBinder := volumebinder.NewVolumeBinder(handle.ClientSet(), nodeInformer, pvcInformer, pvInformer, storageClassInformer, time.Duration(*args.BindTimeoutSeconds)*time.Second)
+	// In Filter phrase, pod binding cache is created for the pod and used in
+	// Reserve and PreBind phrases. Pod binding cache will be cleared at
+	// Unreserve and PostBind extension points.  However, if pod fails before
+	// Reserve phras and is deleted from the apiserver later, its pod binding
+	// cache cannot be cleared at plugin extension points. Here we register an
+	// event handler to clear pod binding cache when the pod is deleted to
+	// prevent memory leaking.
+	// TODO(cofyc) Because pod binding cache is used only in current scheduling
+	// cycle, we can share it via framework.CycleState. Then we don't need to
+	// clear pod binding cache.
+	handle.SharedInformerFactory().Core().V1().Pods().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Comment:
This is a temporary solution. I'm going to leverage CycleState to share state between different extension points in a follow-up PR.